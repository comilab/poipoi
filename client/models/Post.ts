import dayjs, { Dayjs } from 'dayjs'

import Image from '~/models/Image'
import Reaction from '~/models/Reaction'

export type ScopeUnion = 'public'|'password'|'private'

export type RatingUnion = null|'nsfw'|'r18'

export default class Post {
  id: number
  userId: number
  title: string
  caption: string
  showImagesList: boolean
  text: string
  scope: ScopeUnion
  password: string|null
  publishStart: Dayjs|null = null
  publishEnd: Dayjs|null = null
  rating: RatingUnion
  pinned: boolean
  showThumbnail: boolean
  denyRobot: boolean|null
  enableReaction: boolean|null
  allowedEmojis: string[]|null
  deniedEmojis: string[]|null
  enableTwitterShare: boolean|null
  createdAt: Dayjs|null = null
  updatedAt: Dayjs|null = null

  url: string
  textCount: number
  actualDenyRobot: boolean
  actualEnableReaction: boolean
  actualAllowedEmojis: string[]
  actualDeniedEmojis: string[]
  actualEnableTwitterShare: boolean
  emojis: string[] = []

  images: Image[] = []
  imagesCount: number

  reactions: Reaction[] = []
  reactionsCount: number

  constructor (data: any) {
    this.id = data.id
    this.userId = data.userId
    this.title = data.title
    this.caption = data.caption
    this.showImagesList = data.showImagesList
    this.text = data.text
    this.scope = data.scope
    this.password = data.password
    if (data.publishStart) {
      this.publishStart = dayjs.utc(data.publishStart).local()
    }
    if (data.publishEnd) {
      this.publishEnd = dayjs.utc(data.publishEnd).local()
    }
    this.rating = data.rating
    this.pinned = data.pinned
    this.showThumbnail = data.showThumbnail
    this.denyRobot = data.denyRobot
    this.enableReaction = data.enableReaction
    this.allowedEmojis = data.allowedEmojis
    this.deniedEmojis = data.deniedEmojis
    this.enableTwitterShare = data.enableTwitterShare
    if (data.createdAt) {
      this.createdAt = dayjs.utc(data.createdAt).local()
    }
    if (data.updatedAt) {
      this.updatedAt = dayjs.utc(data.updatedAt).local()
    }

    this.url = data.url
    this.textCount = data.textCount
    this.actualDenyRobot = data.actualDenyRobot
    this.actualEnableReaction = data.actualEnableReaction
    this.actualAllowedEmojis = data.actualAllowedEmojis
    this.actualDeniedEmojis = data.actualDeniedEmojis
    this.actualEnableTwitterShare = data.actualEnableTwitterShare
    this.emojis = data.emojis

    if (data.images) {
      this.images = data.images.map((image: any) => new Image(image))
    }
    this.imagesCount = data.imagesCount

    if (data.reactions) {
      this.reactions = data.reactions.map((reaction: any) => new Reaction(reaction))
    }
    this.reactionsCount = data.reactionsCount
  }

  get path () {
    return `/posts/${this.id}`
  }

  get textPages () {
    return (this.text || '').split('[newpage]').map(page => page.trim())
  }
}
