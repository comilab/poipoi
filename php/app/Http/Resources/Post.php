<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\JsonResource;

class Post extends JsonResource
{
    /**
     * Transform the resource into an array.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return array
     */
    public function toArray($request)
    {
        return [
            'id' => $this->id,
            'userId' => $this->user_id,
            'title' => $this->title,
            'caption' => $this->caption,
            'showImagesList' => $this->show_images_list,
            'text' => $this->text,
            'scope' => $this->scope,
            'password' => $this->when(auth()->check(), $this->password),
            'publishStart' => $this->publish_start,
            'publishEnd' => $this->publish_end,
            'rating' => $this->rating,
            'pinned' => $this->pinned,
            'showThumbnail' => $this->show_thumbnail,
            'denyRobot' => $this->deny_robot,
            'enableReaction' => $this->enable_reaction,
            'allowedEmojis' => $this->allowed_emojis,
            'deniedEmojis' => $this->denied_emojis,
            'enableTwitterShare' => $this->enable_twitter_share,
            'createdAt' => $this->created_at,
            'updatedAt' => $this->updated_at,

            'url' => $this->url,
            'textCount' => $this->text_count,
            'actualDenyRobot' => $this->actual_deny_robot,
            'actualEnableReaction' => $this->actual_enable_reaction,
            'actualAllowedEmojis' => $this->actual_allowed_emojis,
            'actualDeniedEmojis' => $this->actual_denied_emojis,
            'actualEnableTwitterShare' => $this->actual_enable_twitter_share,
            'emojis' => $this->emojis,

            'images' => Image::collection($this->whenLoaded('images')),
            'imagesCount' => (int) $this->images_count,

            'reactions' => Reaction::collection($this->when(
                $this->is_mine || $this->actual_enable_reaction,
                $this->whenLoaded('reactions')
            )),
            'reactionsCount' => $this->when(
                $this->is_mine || $this->actual_enable_reaction,
                (int) $this->reactions_count,
            ),
        ];
    }
}
