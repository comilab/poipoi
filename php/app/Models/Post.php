<?php

namespace App\Models;

use App\Services\Emoji;
use App\Services\Setting as SettingService;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Carbon;
use Illuminate\Support\Collection;
use Illuminate\Support\Facades\Storage;

class Post extends Model
{
    use HasFactory;

    protected $fillable = [
        'title',
        'caption',
        'show_images_list',
        'text',
        'scope',
        'password',
        'publish_start',
        'publish_end',
        'rating',
        'pinned',
        'show_thumbnail',
        'deny_robot',
        'enable_reaction',
        'allowed_emojis',
        'denied_emojis',
        'enable_twitter_share',
    ];

    protected $casts = [
        'user_id' => 'integer',
        'show_images_list' => 'boolean',
        'publish_start' => 'datetime',
        'publish_end' => 'datetime',
        'pinned' => 'boolean',
        'show_thumbnail' => 'boolean',
        'deny_robot' => 'boolean',
        'enable_reaction' => 'boolean',
        'allowed_emojis' => 'collection',
        'denied_emojis' => 'collection',
        'enable_twitter_share' => 'boolean',
    ];

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function images()
    {
        return $this->hasMany(Image::class)->orderBy('weight');
    }

    public function reactions()
    {
        return $this->hasMany(Reaction::class)->latest();
    }

    public function tags()
    {
        return $this->belongsToMany(Tag::class);
    }

    public function getUrlAttribute(): string
    {
        return url("/posts/{$this->id}");
    }

    public function getTextCountAttribute(): int
    {
        return mb_strlen($this->text);
    }

    public function getActualDenyRobotAttribute(): bool
    {
        if ($this->deny_robot !== null) {
            return $this->deny_robot;
        }

        $denyRobotScope = app(SettingService::class)->get('postDefault.denyRobotScope');

        return in_array($this->scope, $denyRobotScope, true) || in_array($this->rating, $denyRobotScope, true);
    }

    public function getActualEnableReactionAttribute(): bool
    {
        if ($this->enable_reaction !== null) {
            return $this->enable_reaction;
        }

        return app(SettingService::class)->get('postDefault.enableReaction');
    }

    public function getActualAllowedEmojisAttribute(): Collection
    {
        if ($this->allowed_emojis !== null) {
            return $this->allowed_emojis;
        }

        return collect(app(SettingService::class)->get('postDefault.allowedEmojis'));
    }

    public function getActualDeniedEmojisAttribute(): Collection
    {
        if ($this->denied_emojis !== null) {
            return $this->denied_emojis;
        }

        return collect(app(SettingService::class)->get('postDefault.deniedEmojis'));
    }

    public function getEmojisAttribute(): Collection
    {
        if ($this->actual_allowed_emojis->isNotEmpty() && $this->actual_denied_emojis->isNotEmpty()) {
            return $this->actual_allowed_emojis->reject(function ($emoji) {
                return $this->actual_denied_emojis->contains($emoji);
            });
        } else if ($this->actual_allowed_emojis->isNotEmpty()) {
            return $this->actual_allowed_emojis;
        } else if ($this->actual_denied_emojis->isNotEmpty()) {
            return app(Emoji::class)->all()->reject(function ($emoji) {
                return $this->actual_denied_emojis->contains($emoji);
            });
        } else {
            return collect([]);
        }
    }

    public function getActualEnableTwitterShareAttribute(): bool
    {
        if ($this->enable_twitter_share !== null) {
            return $this->enable_twitter_share;
        }

        return app(SettingService::class)->get('postDefault.enableTwitterShare');
    }

    public function getIsMineAttribute(): bool
    {
        return auth()->check() && (auth()->user()->id === $this->user_id);
    }

    public function getIsPublicAttribute(): bool
    {
        if ($this->scope === 'private') {
            return false;
        }

        if ($this->publish_start && Carbon::now()->isBefore($this->publish_start)) {
            return false;
        }
        if ($this->publish_end && Carbon::now()->isAfter($this->publish_end)) {
            return false;
        }

        return true;
    }

    public function getImagesDirAttribute(): string
    {
        return "public/images/{$this->id}";
    }

    public function scopePublic($query)
    {
        return $query
            ->where('scope', '!=', 'private')
            ->where(function ($query) {
                $query
                    ->orWhere('publish_start', null)
                    ->orWhere('publish_start', '<=', Carbon::now());
            })
            ->where(function ($query) {
                $query
                    ->orWhere('publish_end', null)
                    ->orWhere('publish_end', '>=', Carbon::now());
            });
    }

    public function scopeSearch(Builder $query, $keyword)
    {
        $keywords = preg_split('/[\s]/', $keyword);
        $query->where(function ($query) use ($keywords) {
            foreach ($keywords as $keyword) {
                if ($keyword[0] === '#') {
                    $tag = substr($keyword, 1);
                    $query->whereHas('tags', function (Builder $query) use ($tag) {
                        $query->where('name', $tag);
                    });
                } else {
                    $escaped = str_replace(['\\', '%', '_'], ['\\\\', '\%', '\_'], $keyword);
                    $query->whereRaw("caption like ? escape '\'", ["%{$escaped}%"]);
                }
            }
        });
    }

    public function syncTagsFromCaption()
    {
        preg_match_all('/#(\S+)/', $this->caption, $matches);

        $tags = collect($matches[1])
            ->map(function ($tagName) {
                return Tag::firstOrCreate(['name' => $tagName]);
            });

        $this->tags()->sync($tags->pluck('id'));

        resolve(Tag::class)->clean();

        return $this;
    }

    public function makeImagesDir()
    {
        Storage::makeDirectory($this->images_dir);
    }

    public function deleteImagesDir()
    {
        Storage::deleteDirectory($this->images_dir);
    }
}
