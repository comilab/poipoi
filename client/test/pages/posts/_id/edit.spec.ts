import { render } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import * as NuxtCompositionApi from 'nuxt-composition-api'
import { ref, reactive } from '@vue/composition-api'

import PostsEdit from '~/pages/posts/_id/edit.vue'
import usePostsApi from '~/composables/use-posts-api'
import postFactory from '~/test/factories/post'
import useMockedComponent from '~/test/stubs/MockedComponent'
import Post from '~/models/Post'

jest.mock('~/composables/use-posts-api')

describe('pages/posts/_id/edit', () => {
  let options: any
  let contextMock: any
  let fetchMock: any
  let postsApiMock: any

  beforeEach(() => {
    options = {
      stubs: {
        FontAwesomeIcon: true,
        NuxtLink: true
      }
    }

    contextMock = {
      redirect: jest.fn(),
      params: ref({
        id: 1
      })
    }
    jest.spyOn(NuxtCompositionApi, 'useContext').mockReturnValue(contextMock)

    fetchMock = {
      fetch: jest.fn(),
      fetchState: reactive({
        pending: false
      })
    }
    jest.spyOn(NuxtCompositionApi, 'useFetch').mockReturnValue(fetchMock)

    postsApiMock = {
      show: jest.fn(),
      post: ref<Post>(postFactory.build())
    }
    mocked(usePostsApi).mockReturnValue(postsApiMock)
  })

  test('renders correctly', () => {
    options.stubs.AppFormPost = true
    const { container } = render(PostsEdit, options)

    expect(container).toMatchSnapshot()
  })

  describe('fetch', () => {
    test('showが実行される', async () => {
      options.stubs.AppFormPost = true
      render(PostsEdit, options)

      expect(NuxtCompositionApi.useFetch).toBeCalled()

      const callback = mocked(NuxtCompositionApi.useFetch).mock.calls[0][0] as Function
      await callback()

      expect(postsApiMock.show).toBeCalledWith(1)
    })
  })

  test('onSubmitted', () => {
    const { component, emitsList } = useMockedComponent({
      emits: ['submitted']
    })
    options.stubs.AppFormPost = component
    render(PostsEdit, options)

    emitsList[0].submitted()

    expect(contextMock.redirect).toBeCalledWith(postsApiMock.post.value.path)
  })
})
