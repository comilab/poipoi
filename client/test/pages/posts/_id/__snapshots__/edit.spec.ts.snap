// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/posts/_id/edit renders correctly 1`] = `
<div>
  <div>
    <appformpost-stub
      post="[object Object]"
    />
  </div>
</div>
`;
